### PR TITLE
Add typings

### PR DESCRIPTION
**Pull request by bryanforbes**
This PR adds typings for use with type checkers like `mypy`. Overall, this was fairly easy but two errors have to be suppressed with a `# type: ignore` comment:

* `remove_signal_handler()` returns a `bool` and typeshed has this method returning `None` in `AbstractEventLoop`. This seems to be an error in typeshed since the [docs](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.remove_signal_handler) state it should return a bool.
* `create_connection()` in Python 3.8+ accepts two new arguments (`happy_eyeballs_delay` and `interleave`) and since `uvloop.Loop.create_connection()` doesn't allow those arguments, the two methods are incompatible. This may be unavoidable.

References #358

### CODE REVIEW

File: uvloop/loop.pyi
Comment (1st1) [inline uvloop/loop.pyi:110]: I'm curious why we need this as opposed to simply using `asyncio.AbstractServer`?

Comment (bryanforbes) [reply uvloop/loop.pyi:110]: `asyncio.AbstractServer` is defined as follows in typeshed:

```python
class AbstractServer:
    sockets: Optional[List[socket]]
    def close(self) -> None: ...
    if sys.version_info >= (3, 7):
        async def __aenter__(self: _T) -> _T: ...
        async def __aexit__(self, *exc: Any) -> None: ...
        def get_loop(self) -> AbstractEventLoop: ...
        def is_serving(self) -> bool: ...
        async def start_serving(self) -> None: ...
        async def serve_forever(self) -> None: ...
    async def wait_closed(self) -> None: ...
```

Since `uvloop` works on Python 3.5+ and `Server` defines those methods for all versions, `asyncio.AbstractServer` will be wrong for `uvloop` on Python 3.5 and 3.6.

Comment (1st1) [reply uvloop/loop.pyi:110]: I'll be dropping support for 3.5 and 3.6. 3.5 is no longer supported by Python core devs; 3.6 is in security fixes only mode. And with 3.9 now out, uvloop will be 3.7+ only in its next release, supporting three Python versions: 3.7, 3.8, and 3.9.

Comment (bryanforbes) [reply uvloop/loop.pyi:110]: That's good to know! I'll go back over this PR with that in mind.

Comment (bryanforbes) [reply uvloop/loop.pyi:110]: One thing to note: I think we'll still need `_Server` because `AbstractServer.sockets` is `Optional[List[Socket]]` and `Server.sockets` is `List[Socket]`.

Comment (1st1) [reply uvloop/loop.pyi:110]: I think we should stick to AbstractServer typing here too. That makes it easy for us to fix our `Server.sockets` property to follow the spec.

Comment (bryanforbes) [reply uvloop/loop.pyi:110]: Some of these others I added while I was navigating the source. I'll make sure I validate all of these are needed when I go over this again.

Comment (1st1) [reply uvloop/loop.pyi:110]: Yes, thanks. And we can always fix uvloop code that doesn't conform to asyncio APIs.

Comment (bryanforbes) [reply uvloop/loop.pyi]: Sorry that it's taken so long to get back to this. I've updated the PR based on your feedback and rebased on master. Unfortunately, the PR checks are failing prematurely.

**Comment by 1st1** [comment]
@fantix We can probably merge this one too, please review.

**Comment by bryanforbes** [comment]
I can rebase this on master if it would make it easier to review. Just let me know.

**Comment by fantix** [comment]
@bryanforbes Yes, please! Thank you!

**Comment by bryanforbes** [comment]
I have a couple of small edits to make before you merge

**Review by fantix — APPROVED** [review]
LGTM!

**Comment by bryanforbes** [comment]
@fantix I committed and pushed the edits. Let me know if you have questions about them. One thing to note: the two things I pointed out in the description of this PR are still issues. Once those methods are changed, the `# type: ignore[misc]` can be removed from `__init__.py`.

**Comment by fantix** [comment]
> Once those methods are changed, the `# type: ignore[misc]` can be removed from `__init__.py`.

Sounds good 👍

**Comment by 1st1** [comment]
Thanks @bryanforbes!

**Comment by bryanforbes** [comment]
Not a problem! If you ever have any typing questions, feel free to reach out to me.

**Comment by 1st1** [comment]
@bryanforbes thanks! If you have time please take a look at https://github.com/MagicStack/immutables/pull/54 -- another our package.

**Pull request merged**
